### ISSUE TITLE
NVDA does not read lines of TRichViewEdit text control in Delphi application

### ISSUE THREAD

**Issue by jimbr32** [nvaccess/nvda#8166]
### Steps to reproduce:
Start my application and load text into the RichViewEdit text control.
Arrow up and down in the text.

The attached zip contains a sample app I'm using to duplicate the problem.
[SampleRVApp.zip](https://github.com/nvaccess/nvda/files/1900350/SampleRVApp.zip)

### Expected behavior:
NVDA should speak the current line

### Actual behavior:
NVDA speaks the text 'blank'

### System configuration:
NVDA version: 2018.1.1

NVDA Installed or portable: Installed

Other information:
Our customers are seeing this problem on actual machines, I am duplicating issue in a VM (VM Workstation)

Windows version:
I am duplicating the issue on Windows 10 Pro, version 1709, Build 16299.371

Name and version of other software in use when reproducing the issue:

### Other questions:

Does the issue still occur after restarting your PC?
Yes

Have you tried any other versions of NVDA?
No

**Comment by feerrenrut** [comment]
What have you done to investigate this issue?
Do you get different behaviour with windows narrator?
I would suggest using a tool like `inspect.exe` available in a VisualStudio installation under a path like `C:\Program Files (x86)\Windows Kits\10\bin\10.0.15063.0\x86` to see what is being exposed.
You haven't provided the source for this sample app, which makes it pretty hard to suggest anything.

**Comment by jimbr32** [comment]
Here is more information.

I have tested my application with both JAWS and Windows Narrator.  When I
arrow up and down in the text, JAWS does speak the current line, and speaks
the current character if I arrow left and right.  Narrator only speaks
which arrow key was pressed.  "Up Arrow", "Down Arrow", etc.  Narrator does
not read the current line / letter.

In the previous version of my application I used a component called
TPlusMemoU, and switch to RichViewEdit in the current version.  NVDA is
able to read the current line as I arrow up and down in the PlusMemoU
control, and the current character as I arrow left and right.  I attached a
zip containing how the two controls look to the Inspect utility.  The zip
contains msaa and ui automation (uia) views of both RichViewEdit (RVE) and
PlusMemoU (PMU) controls.

I received a reply from the RichViewEdit author.  Accessibility API's are
not currently implememted, but he did say this:

*TRichViewEdit processes Edit and RichEdit messages, including EM_GETSEL,
EM_SETSEL, EM_GETTEXTRANGE*
*So treating it like a multiline editor control must work.*
*If they need some additional message to be implemented, I can do it, it
should not be a problem.*

As far as source, the source code for the sample app is just a form with a
richviewedit control dropped on it.  The only code in the app is the code
used to write a few lines of text to the control when you start the app.  I
am not able to supply source for the RichViewEdit component.

Does any of this help?

Jim

**Comment by leonardder** [comment]
It looks like that when I patch NVDA in such a way that it uses EM_GETTEXTRANGE, the results are still not as expected.

**Comment by feerrenrut** [comment]
Re: source code.
Personally, I don't like to run executables from unknown sources (no offense). Even within a VM. I would much prefer to view the source and build it myself. Firstly this allows me to inspect that the example is safe to run. This might seem paranoid, but given the hassle of restoring a system I'd prefer to be safe than sorry . Secondly, it allows me to investigate the components / libraries used and understand what I am expected to find when I run the application.

>  The only code in the app is the code used to write a few lines of text to the control when you start the app

What is this text? So we know what we are looking for.

**Comment by jimbr32** [comment]
Here is the source code for the sample app.  I compiled from this source using Delphi 10 Seattle.

[Sample RVE Source.zip](https://github.com/nvaccess/nvda/files/1925037/Sample.RVE.Source.zip)

To build the app you will also need to download and install a trial version of the TRichView component from https://www.trichview.com/download/

**Comment by jimbr32** [comment]
One of our customers came up with a workaround that solves this problem.
Here are the steps:

1. If my application is running, exit both NVDA and my application, and
restart NVDA.

2. In the NVDA python console, enter this:

import NVDAObjects.window;del
NVDAObjects.window.windowClassMap['TRichViewEdit']

3. Start my application.

NVDA now reads the current line as I arrow up and down in the RichViewEdit
control, and the current character if I arrow left and right.  I verified
this works with my actual application and the sample app I posted earlier.

Is this a problem with the RichViewEdit mapping
in NVDAObjects/window/__init__.py ?

Jim Bretti
NextUp.com <http://nextup.com>

**Comment by leonardder** [comment]
I believe this workaround will cause the display model to be used, which is always a last resort to get information from a text control. If the control supports the Rich Edit API, that should be prefered. I might be able to investigate this problem further later this or next week.

**Comment by jimbr32** [comment]
I tried this from the python console and it seems to work:

import NVDAObjects.window
NVDAObjects.window.windowClassMap['TRichViewEdit']='RichEdit50W'

I also tried RichEdit and RichEdit20 with no luck, RichEdit50W seems to be
working pretty well so far

**Comment by leonardder** [comment]
[**@jimbr32**](https://github.com/jimbr32) commented on [20 Apr 2018, 21:32 CEST](https://github.com/nvaccess/nvda/issues/8166#issuecomment-383199008 "2018-04-20T19:32:34Z - Replied by Github Reply Comments"):
> I tried this from the python console and it seems to work:
> 
> import NVDAObjects.window
> NVDAObjects.window.windowClassMap['TRichViewEdit']='RichEdit50W'
> 
> I also tried RichEdit and RichEdit20 with no luck, RichEdit50W seems to be
> working pretty well so far

I'm afraid that the only reason it works is that RichEdit50W is a nonexistent window class in NVDA's map, it should be RICHEDIT50W. In your case, NVDA also falls back to the display model, and indeed, that works somehow, though it doesn't use the underlying API.

I've figured out that TRichViewEdit neither supports EM_GETTEXTLENGTHEX or WM_GETTEXTLENGTH to get the text length.

Are you really required to use this edit control? I know that TMemo works quite well.

**Comment by jimbr32** [comment]
Thanks for the help Leonard, I think I understand this better now.  I
looked at the messaging closer using Spy, using the default class mapping
of TRichViewEdit to TEdit.  I noticed that WM_GETTEXTLENGTH was not being
handled, along with 3 or 4 other messages.  I was able to add code in my
test app to handle these messages, and I think I have it working.  I should
be able to work with the component author from here

Thanks again for the help.

**Comment by leonardder** [comment]
You're welcome. Feel free to get in touch if you have any additional questions.

I'll label this as "needs external fix". Could you please get in touch as soon as the issues are resolved in the TRichViewComponent itself?

**Comment by jimbr32** [comment]
Yes, I'll post here when the problems are resolved.

**Comment by jimbr32** [comment]
The problems are fixed in TRichView component version 17.4, the author added support for the missing edit control messages.  I'll go ahead and close the issue, I appreciate all the help.
